src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ILScanner.cs

```diff
@@ -674,6 +679,75 @@ public override TypeSystemException GetCompilationError(MethodDesc method)
                 => _importationErrors.TryGetValue(method, out var exception) ? exception : null;
         }
 
+        private sealed class ScannedInlinedThreadStatics : InlinedThreadStatics
+        {
+            private readonly List<MetadataType> _types;
+            private readonly Dictionary<MetadataType, int> _offsets;
+            private readonly int _size;
+
+            internal override bool IsComputed() => true;
+            internal override List<MetadataType> GetTypes() => _types;
+            internal override Dictionary<MetadataType, int> GetOffsets() => _offsets;
+            internal override int GetSize() => _size;
+
+            public ScannedInlinedThreadStatics(NodeFactory factory, ImmutableArray<DependencyNodeCore<NodeFactory>> markedNodes)
+            {
+                List<ThreadStaticsNode> threadStaticNodes = new List<ThreadStaticsNode>();
+                foreach (var markedNode in markedNodes)
+                {
+                    if (markedNode is ThreadStaticsNode threadStaticNode)
+                    {
+                        threadStaticNodes.Add(threadStaticNode);
+                    }
+                }
+
+                // skip MT pointer
+                int nextDataOffset = factory.Target.PointerSize;
+
+                List<MetadataType> types = new List<MetadataType>();
+                Dictionary<MetadataType, int> offsets = new Dictionary<MetadataType, int>();
+
+                if (threadStaticNodes.Count > 0)
+                {
+                    threadStaticNodes.Sort(CompilerComparer.Instance);
+                    for (int i = 0; i < threadStaticNodes.Count; i++)
+                    {
+                        ThreadStaticsNode threadStaticNode = threadStaticNodes[i];
+                        MetadataType t = threadStaticNode.Type;
+                        // REVIEW: how to filter these more precisely?
+                        //       List<int> is ok, but List<object> is not.
+                        if (t.HasInstantiation)
+                            continue;
```

Comment:
The issue here is that from outside `SomeType<object>.t_field` might be considered inlineable. It looks like we treat instantiations with `object` in a special way - as a one kind of instantiation that does not need dynamic threadstatic storage. It would work correclty as long as all accesses are from outside of the type.

But from inside of `SomeType<T>` accessing `[threadstatic] T t_field` would be a dictionary lookup. The lookup can figure where the field is stored and get the storage for it at run time, but we would have specified at compile time a wrong field offset that expects dynamic lookup and no inlining. 

I think this is only a problem when `T` is substituted for `object`. I wonder if there is a way to filter out the affected cases more precisely. 
This should stop being an issue when storage is back to the per-type plan.

Reply:
I think it should be `type.ConvertToCanonForm(CanonicalFormKind.Specific) != type` - similar to the othe place.